### PR TITLE
APPSI-799

### PR DESCRIPTION
**Pull request by cmillar-r7**
Add proxy connection. 
Add server logs debugging.
Update readme/versioning.

### Description

Added options to allow connection via proxy in plugin config page.
Added option to enable server log output for debugging purposes. 
Modified error message output in logging page.

### Motivation and Context

[https://issues.corp.rapid7.com/browse/APPSI-799](https://issues.corp.rapid7.com/browse/APPSI-799)
Proxy functionality and greater debugging provision 

### How Has This Been Tested?

Can kick off scan using proxy settings 
	Scan results in apses

Can enable debugging 
	Debugging output shown in server logs 

Can not enter invalid appsec settings:
	- App name
	- Scan config name 
	- Api 
		Appropriate error is shown 

### Types of changes

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to change)
New feature 

### Checklist:

- [ ] I have updated the documentation accordingly (if changes are required).
- [ ] I have tested the changes locally.

### CODE REVIEW

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java [hunk 1]
```diff
@@ -9,6 +9,7 @@
 import com.rapid7.ias.bamboo.util.UtilityLogger;
 import com.rapid7.ias.client.model.*;
 import com.squareup.okhttp.Call;
+import scala.util.parsing.combinator.testing.Str;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java]: Pretty sure you didn't mean to import this?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java [hunk 2]
```diff
@@ -31,12 +32,12 @@ public class InsightAppSecHelper {
...
 
-    public InsightAppSecHelper(String region, String apiKey, UtilityLogger logger) {
+    public InsightAppSecHelper(String region, String apiKey, UtilityLogger logger, String proxyHost, String proxyPort, String debugging) {
         this.logger = logger;
 
         // Build generic client with API key set
-        ApiClient client = new ApiClient();
-        client.setBasePath("https://" + region + ".api.insight.rapid7.com/ias/v1");
+        ApiClient client = new ApiClient(proxyHost, proxyPort).setDebugging(Boolean.parseBoolean(debugging));
+                client.setBasePath("https://" + region + ".api.insight.rapid7.com/ias/v1");
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java]: Shouldn't be extra indentation here?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java [hunk 3]
```diff
@@ -31,12 +32,12 @@ public class InsightAppSecHelper {
...
     public VulnerabilitiesApi vulnerabilitiesApi;
 
-    public InsightAppSecHelper(String region, String apiKey, UtilityLogger logger) {
+    public InsightAppSecHelper(String region, String apiKey, UtilityLogger logger, String proxyHost, String proxyPort, String debugging) {
         this.logger = logger;
 
         // Build generic client with API key set
-        ApiClient client = new ApiClient();
-        client.setBasePath("https://" + region + ".api.insight.rapid7.com/ias/v1");
+        ApiClient client = new ApiClient(proxyHost, proxyPort).setDebugging(Boolean.parseBoolean(debugging));
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecHelper.java:40]: Will debugging ever be null? Will it only ever be "true" or "false"?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTask.java [hunk 4]
```diff
@@ -54,6 +58,11 @@ public TaskResult execute(CommonTaskContext taskContext) throws TaskException {
         appName = configMap.get(APP_NAME);
         scanConfigName = configMap.get(SCAN_CONFIG_NAME);
 
+        proxyHost = configMap.get(PROXY_HOST);
+        proxyPort = configMap.get(PROXY_PORT);
+
+        debugging = configMap.get(debugging);
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTask.java]: Why don't you have a constant for the debugging property name, like with PROXY_HOST and PROXY_PORT?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package com.rapid7.ias.bamboo.impl;
 
+import aQute.bnd.http.HttpClient;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: Did you mean to import this http client? - Check for unused imports and remove them

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 6]
```diff
@@ -92,39 +106,56 @@ public Map<String,String> generateTaskConfigMap(@NotNull ActionParametersMap par
...
-        }else if (StringUtils.isEmpty(regionValue)){
+        } else if (StringUtils.isEmpty(regionValue)) {
             errorCollection.addError(SELECTED_REGION, i18nBean.getText("error.region"));
-        }else if(StringUtils.isEmpty(appNameValue)){
+        } else if (StringUtils.isEmpty(appNameValue)) {
             errorCollection.addError(APP_NAME, i18nBean.getText("error.appName"));
-        }else if(StringUtils.isEmpty(scanConfigNameValue)){
+        } else if (StringUtils.isEmpty(scanConfigNameValue)) {
             errorCollection.addError(SCAN_CONFIG_NAME, i18nBean.getText("error.scanConfigName"));
+        } else if (StringUtils.isNotEmpty(proxyHostValue) && !proxyHostValue.matches("^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$")) {
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: Can they use a dns name rather than IP for the proxy host?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 7]
```diff
@@ -137,6 +168,13 @@ public void validate(@NotNull ActionParametersMap params,
         }
     }
 
+    private boolean checkProxyConnection(String proxyHostValue, String proxyPortValue) {
+        ApiClient apiClient = new ApiClient(proxyHostValue, proxyPortValue);
+        apiClient.setBasePath("https://www.google.com");
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: Maybe try rapid7.com instead?

File: src/main/java/com/rapid7/ias/client/ApiClient.java [hunk 8]
```diff
@@ -92,6 +99,27 @@ public ApiClient() {
...
 
+    private void addProxy(String proxyHost, String proxyPort) {
+        proxyAddr = new InetSocketAddress(proxyHost, Integer.parseInt(proxyPort));
+
+        proxy = new Proxy(Proxy.Type.HTTP, proxyAddr);
+        httpClient.setProxy(proxy);
+    }
+
+    public boolean checkProxyConnection() {
+      Request request = new Request.Builder()
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/client/ApiClient.java]: Minor nitpick but make sure the indentation is consistent in this method.

File: src/main/java/com/rapid7/ias/client/ApiClient.java [hunk 9]
```diff
@@ -71,13 +73,18 @@ public class ApiClient {
     private JSON json;
 
     private HttpLoggingInterceptor loggingInterceptor;
+    InetSocketAddress proxyAddr;
+    private Proxy proxy;
 
     /*
      * Constructor for ApiClient
      */
-    public ApiClient() {
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/client/ApiClient.java:78]: Might be worth keeping in an empty constructor also, so tests don't have to pass in a proxy or null

File: src/main/resources/com/rapid7/ias/bamboo/i18n.properties [hunk 10]
```diff
@@ -24,12 +27,16 @@ error.scanConfigName=You did not define a Scan Config
 error.authFailure=Authentication failure while verifying configuration, please ensure the InsightAppec API Key is valid
 error.invalidAppConfig=App does not exist, please confirm the spelling
 error.invalidScanConfig=Scan Config does not exist, please confirm the spelling and verify it exists to specified App
+error.proxyHost=Enter in ip format
```
Comment (ckearney-r7) [inline src/main/resources/com/rapid7/ias/bamboo/i18n.properties]: Id just remove the regex validation if they can only enter in the IP

File: src/main/resources/com/rapid7/ias/bamboo/i18n.properties [hunk 11]
```diff
@@ -24,12 +27,16 @@ error.scanConfigName=You did not define a Scan Config
...
+error.proxyHost=Enter in ip format 
+error.proxyPort=Enter numeric port
+error.proxyConnection=Check proxy connection settings
 
 section.connection=Connection Details
 section.action=Task Advancement
 section.status=Task Options
 section.gating=Scan Gating
 section.gating.details=For each scan, it is possible to enforce a Vulnerability Query gate. Any vulnerabilities matching the filter will fail the task.
+section.debug=Debugging
```
Comment (ckearney-r7) [inline src/main/resources/com/rapid7/ias/bamboo/i18n.properties]: Maybe "Debug Logging" might make this clearer?

**Review by ckearney-r7 — COMMENTED** [review]
Did you mean to create an empty file called "atlas-install-plugin"?

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 12]
```diff
@@ -11,6 +11,13 @@
 import com.atlassian.bamboo.utils.i18n.I18nBean;
 import com.atlassian.bamboo.utils.i18n.I18nBeanFactory;
 import static com.atlassian.bamboo.credentials.UsernamePasswordCredentialType.CFG_PASSWORD;
+
+import com.rapid7.ias.client.ApiClient;
+import com.squareup.okhttp.Call;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: Should remove all these unused imports

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 13]
```diff
@@ -21,6 +28,9 @@
 import org.apache.commons.lang.StringUtils;
 import org.apache.log4j.Logger;
 
+import java.io.IOException;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: And these

File: src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java [hunk 14]
```diff
@@ -92,39 +105,54 @@ public Map<String,String> generateTaskConfigMap(@NotNull ActionParametersMap par
...
+        } else if (StringUtils.isEmpty(scanConfigNameValue)) {
             errorCollection.addError(SCAN_CONFIG_NAME, i18nBean.getText("error.scanConfigName"));
+        } else if (StringUtils.isNotEmpty(proxyPortValue) && !StringUtils.isNumeric(proxyPortValue)) {
+            errorCollection.addError(PROXY_PORT, i18nBean.getText("error.proxyPort"));
         }else{
             CredentialsData cred = credentialsAccessor.getCredentials(Long.parseLong(credentialId));
 
-            InsightAppSecHelper iasHelper = new InsightAppSecHelper(regionValue, cred.getConfiguration().get(CFG_PASSWORD), logger);
+            if (proxyHostValue.isEmpty()) {
+                proxyHostValue = proxyPortValue = null;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/bamboo/impl/InsightAppSecScanTaskConfigurator.java]: Can we do this on multiple lines

proxyHostValue = null;
proxyPortValue = null;

This is valid, but not something ive seen done very often

File: src/main/java/com/rapid7/ias/client/ApiClient.java [hunk 15]
```diff
@@ -71,13 +73,22 @@ public class ApiClient {
     private JSON json;
 
     private HttpLoggingInterceptor loggingInterceptor;
+    InetSocketAddress proxyAddr;
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/client/ApiClient.java]: These can be local variables in addProxy

File: src/main/java/com/rapid7/ias/client/Configuration.java [hunk 16]
```diff
@@ -15,7 +15,7 @@
 
 @javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaClientCodegen", date = "2018-11-27T23:32:49.618-05:00")
 public class Configuration {
-    private static ApiClient defaultApiClient = new ApiClient();
+    private static ApiClient defaultApiClient = new ApiClient(null,null);
```
Comment (ckearney-r7) [inline src/main/java/com/rapid7/ias/client/Configuration.java]: Can probably just use the default constructor that you added rather than passing null?

**Review by ckearney-r7 — APPROVED** [review]

**Review by trobinson-r7 — APPROVED** [review]
